odl/discr/tensor_ops.py

```diff
@@ -0,0 +1,610 @@
+﻿# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Operators defined for tensor fields."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+from builtins import super
+
+import numpy as np
+
+from odl.operator.operator import Operator
+from odl.set.sets import RealNumbers, ComplexNumbers
+from odl.space.fspace import FunctionSpace
+from odl.space.pspace import ProductSpace
+
+
+__all__ = ('PointwiseOperator', 'PointwiseNorm', 'PointwiseInner')
+
+
+class PointwiseOperator(Operator):
+
+    """Abstract operator for point-wise tensor field manipulations.
+
+    A point-wise operator acts on a space of vector or tensor fields,
+    i.e. a power space ``X^d`` of a discretized function space ``X``.
+    Its range is the power space ``X^k`` with a possibly different
+    number ``k`` of components.
+
+    For example, if ``X`` is a `DiscreteLp` space, then
+    ``ProductSpace(X, d)`` is a valid domain for any positive integer
+    ``d``.
+
+    Currently, only vector fields, i.e. one-dimensional products of
+    ``X``, are supported.
+
+    See also
+    --------
+    ProductSpace
+    """
+
+    def __init__(self, domain, range, linear=False):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        domain, range : {`ProductSpace`, `Discretization`}
+            Spaces of vector fields between which the operator maps.
+            They have to be either power spaces of the same base space
+            ``X`` or the base space itself (only one of them).
+        linear : `bool`, optional
+            If `True`, assume that the operator is linear.
+        """
+        if isinstance(domain, ProductSpace):
+            if not domain.is_power_space:
+                raise TypeError('domain {!r} is not a power space.'
+                                ''.format(domain))
+
+            if not isinstance(domain[0].uspace, FunctionSpace):
+                raise TypeError('domain base space {!r} is not a function '
+                                'space discretization.'.format(domain[0]))
+
+            dom_base = domain[0]
```

Comment:
We actually accept 0 dimensional productspaces, in which case this fails. Dunno if we want to check for that though, its mostly a hack to enable things like:
```python
# space is 3 dimensional
for x in space[3:]: print(x)
```